File: Software/workspace/SimTankBot/src/Commands/DrivePath.cpp
#include "DrivePath.h"

#define I2M(x) x*0.0254

//static Waypoint p1 = { -4, -1, d2r(45) };      // Waypoint @ x=-4, y=-1, exit angle=45 degrees
//static Waypoint p2 = { -1, 2, 0 };             // Waypoint @ x=-1, y= 2, exit angle= 0 radians
//static Waypoint p3 = {  2, 4, 0 };             // Waypoint @ x= 2, y= 4, exit angle= 0 radians

#define POINT_LENGTH 4
static Waypoint points[POINT_LENGTH];

static Waypoint p0 = { 0, 0, 0 };
static Waypoint p1 = { 2, 0, 0 };
static Waypoint p2 = { 4, 4, d2r(45)};
static Waypoint p3 = { 8, 8, 0 };

static TrajectoryCandidate candidate;

//#define PRINT_PATH
//#define PLOT_PATH
#define PLOT_TRAJECTORY

//#define USE_GYRO

#define TIME_STEP 0.02
#define MAX_VEL 2.1 //2.75
#define MAX_ACC 2.0
#define MAX_JRK 2.0
#define KP 1.4
#define KI 0.0
#define KD 0.0
#define KV 1.0/MAX_VEL
#define KA 0.0
#define WHEELBASE_WIDTH  0.85 //0.64

static Timer mytimer;
static double runtime;

DrivePath::DrivePath() : config{KP,KI,KD,KV,KA} {
	Requires(driveTrain.get());
	points[0] = p0;
	points[1] = p1;
	points[2] = p2;
	points[3] = p3;

    std::cout << "new DrivePath()"<< std::endl;


	// Prepare the Trajectory for Generation.
	//
	// Arguments:
	// Fit Function:        FIT_HERMITE_CUBIC or FIT_HERMITE_QUINTIC
	// Sample Count:        PATHFINDER_SAMPLES_HIGH (100 000)
	//                      PATHFINDER_SAMPLES_LOW  (10 000)
	//                      PATHFINDER_SAMPLES_FAST (1 000)
	// Time Step:           0.001 Seconds
	// Max Velocity:        15 m/s
	// Max Acceleration:    10 m/s/s
	// Max Jerk:            60 m/s/s/s

	pathfinder_prepare(points, POINT_LENGTH, FIT_HERMITE_CUBIC, PATHFINDER_SAMPLES_FAST, TIME_STEP, MAX_VEL, MAX_ACC, MAX_JRK, &candidate);

	length = candidate.length;

	// Array of Segments (the trajectory points) to store the trajectory in
	trajectory = (Segment*)malloc(length * sizeof(Segment));
	leftTrajectory = (Segment*)malloc(length * sizeof(Segment));
	rightTrajectory = (Segment*)malloc(length * sizeof(Segment));

	// Generate the trajectory
	int result = pathfinder_generate(&candidate, trajectory);
	if (result < 0) {
	    // An error occured
	    printf("Uh-Oh! Trajectory could not be generated!\n");
		cleanup();
	    return;
	}
	runtime=length*TIME_STEP;
	printf("Trajectory length:%d path_length:%d totalLength:%f travel_time:%f\n",
			 candidate.length,candidate.path_length,candidate.totalLength,runtime);
#ifdef PRINT_PATH
	for (int i = 0; i < length; i++) {
	    Segment s = trajectory[i];
	    printf("Time Step: %f\n", s.dt);
	    printf("Coords: (%f, %f)\n", s.x, s.y);
	    printf("Position (Distance): %f\n", s.position);
	    printf("Velocity: %f\n", s.velocity);
	    printf("Acceleration: %f\n", s.acceleration);
	    printf("Jerk (Acceleration per Second): %f\n", s.jerk);
	    printf("Heading (radians): %f\n", s.heading);
	}
#endif
#ifdef PLOT_PATH
	double t;
	for (int i = 0; i < length; i++) {
	    Segment s = trajectory[i];
	    printf("%f %f %f %f %f %f \n", t, s.x, s.y,s.velocity, s.acceleration,s.heading);
	    t+=s.dt;
	}
#endif

    // Generate the Left and Right trajectories of the wheelbase using the
    // originally generated trajectory
	ModifyTank();
    std::cout << "Path generated length="<<length<< std::endl;
#ifdef PLOT_TRAJECTORY
    double tm=0;
	for (int i = 0; i < length; i++) {
		Segment l = leftTrajectory[i];
		Segment r = rightTrajectory[i];
	    //printf("%f %f %f %f %f\n", tm, l.position,r.position,l.velocity,r.velocity);
	    printf("%f %f %f %f %f\n", tm, l.x,l.y,r.x,r.y);

	    tm+=TIME_STEP;
	}
#endif
}

DrivePath::~DrivePath() {
	cleanup();
}

// Called just before this Command runs the first time
void DrivePath::Initialize() {
	leftFollower.finished=0;leftFollower.last_error=0;leftFollower.segment=0;
	rightFollower.finished=0;rightFollower.last_error=0;rightFollower.segment=0;
	std::cout << "DrivePath Started .."<< std::endl;
	driveTrain->Reset();
	mytimer.Start();
	mytimer.Reset();
}

// Called repeatedly when this Command is scheduled to run
//double pathfinder_follow_distance(FollowerConfig c, DistanceFollower *follower, Segment *trajectory, int trajectory_length, double distance) {
#define DEBUG_COMMAND
#define AFACT 20
void DrivePath::Execute() {
	double ld=I2M(driveTrain->GetLeftDistance());
	double rd=I2M(driveTrain->GetRightDistance());
	double l=FollowDistance(&leftFollower,leftTrajectory,ld);
	double r=FollowDistance(&rightFollower,rightTrajectory,rd);
	double lt=leftTrajectory[leftFollower.segment-1].position;
	double rt=rightTrajectory[rightFollower.segment-1].position;
	double ltv=leftTrajectory[leftFollower.segment-1].velocity;
	double rtv=rightTrajectory[rightFollower.segment-1].velocity;
	double lv=I2M(driveTrain->GetLeftVelocity());
	double rv=I2M(driveTrain->GetRightVelocity());
	double rerr=rt-rd;
	double lerr=lt-ld;

	if(leftFollower.finished)
		l=config.kp*lerr;

	if(rightFollower.finished)
		r=config.kp*rerr;
	double gh = driveTrain->GetHeading() ;     // Assuming gyro angle is given in degrees
	double th = r2d(leftFollower.heading);
	th=th>180?360-th:th;
	double herr = th - gh;    // Make sure to bound this from -180 to 180, otherwise you will get super large values
#ifdef USE_GYRO
	double turn = AFACT * (-1.0/180.0) * herr;
#else
	double turn=0;
#endif

#ifdef DEBUG_COMMAND
    printf("%f %f %f %f %f %f %f %f %f %f %f %f\n", mytimer.Get(), ld, lt, rd, rt, gh,th,lerr,rerr,herr,r,l);
    //printf("%f %f %f %f %f %f %f %f %f %f %f %f\n", mytimer.Get(), lv, ltv, rv, rtv, gh,th,ltv-lv,rtv-rv,th-gh,l+turn,r-turn);
#endif
	driveTrain->TankDrive(l+turn,r-turn);
}

#define MAX_ANGLE_ERROR 1   // degrees
#define MAX_POSITION_ERROR 0.5 // meters
// Make this return true when this Command no longer needs to run execute()
bool DrivePath::IsFinished() {
	if(trajectory==NULL)
		return true;
	if(mytimer.Get()-runtime>2)
		return true;

	if(leftFollower.finished && rightFollower.finished){ // end of calculated path
		double ld=I2M(driveTrain->GetLeftDistance());
		double rd=I2M(driveTrain->GetRightDistance());
		double lerr=leftTrajectory[length-1].position-ld;
		double rerr=rightTrajectory[length-1].position-rd;
		if(fabs(lerr>MAX_POSITION_ERROR) ||  fabs(rerr)>MAX_POSITION_ERROR){
			return false;
		}
		return true;
	}
	return false;
}

// Called once after isFinished returns true
void DrivePath::End() {
    std::cout << "DrivePath End" << std::endl;
	mytimer.Stop();
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void DrivePath::Interrupted() {
    std::cout << "DrivePath Interrupted" << std::endl;
}

void DrivePath::cleanup() {
	if(trajectory)
		delete trajectory;
	if(leftTrajectory)
		delete leftTrajectory;
	if(rightTrajectory)
		delete rightTrajectory;
	trajectory=NULL;
	leftTrajectory=NULL;
	rightTrajectory=NULL;
}
void DrivePath::ModifyTank() {
    double w = WHEELBASE_WIDTH / 2;
    int i;
    for (i = 0; i < length; i++) {
        Segment seg = trajectory[i];
        Segment left = seg;
        Segment right = seg;

        double cos_angle = cos(seg.heading);
        double sin_angle = sin(seg.heading);

        left.x = seg.x - (w * sin_angle);
        left.y = seg.y + (w * cos_angle);

        if (i > 0) {
            Segment last = leftTrajectory[i - 1];
            double distance = sqrt(
                (left.x - last.x) * (left.x - last.x)
                + (left.y - last.y) * (left.y - last.y)
            );
            left.position = last.position + distance;
            left.velocity = distance / seg.dt;
            left.acceleration = (left.velocity - last.velocity) / seg.dt;
            left.jerk = (left.acceleration - last.acceleration) / seg.dt;
        }

        right.x = seg.x + (w * sin_angle);
        right.y = seg.y - (w * cos_angle);

        if (i > 0) {
            Segment last = rightTrajectory[i - 1];
            double distance = sqrt(
                (right.x - last.x) * (right.x - last.x)
                + (right.y - last.y) * (right.y - last.y)
            );

            right.position = last.position + distance;
            right.velocity = distance / seg.dt;
            right.acceleration = (right.velocity - last.velocity) / seg.dt;
            right.jerk = (right.acceleration - last.acceleration) / seg.dt;
        }

        leftTrajectory[i] = left;
        rightTrajectory[i] = right;
    }
}
double DrivePath::FollowDistance(DistanceFollower *follower, Segment *trajectory, double distance) {
    int segment = follower->segment;
    if (segment >= length) {
        follower->finished = 1;
    } else {
		follower->finished = 0;
    }
	Segment s = trajectory[segment];
	double error = s.position - distance;
	double calculated_value = config.kp * error +
			config.kd * ((error - follower->last_error) / s.dt) +
							  (config.kv * s.velocity + config.ka * s.acceleration);

	follower->last_error = error;
	follower->heading = s.heading;
	follower->output = calculated_value;
	follower->segment = follower->segment>=length-1?length-1:follower->segment + 1;
	return calculated_value;
}

File: Software/workspace/JSimBot2019/src/org/usfirst/frc/team159/robot/commands/DrivePath.java
package org.usfirst.frc.team159.robot.commands;

import org.usfirst.frc.team159.robot.Robot;

import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.command.Command;
import jaci.pathfinder.Pathfinder;
import jaci.pathfinder.Trajectory;
import jaci.pathfinder.Trajectory.Segment;
import jaci.pathfinder.Waypoint;
import jaci.pathfinder.followers.DistanceFollower;
import jaci.pathfinder.modifiers.TankModifier;
import org.usfirst.frc.team159.robot.subsystems.PlotPath;
import org.usfirst.frc.team159.robot.PathData;
import java.util.ArrayList;
import javax.swing.JFrame;

public class DrivePath extends Command {
	Trajectory trajectory;
	Trajectory leftTrajectory;
	Trajectory rightTrajectory;
	DistanceFollower leftFollower;
	DistanceFollower rightFollower;
	Trajectory.Config config;
	TankModifier modifier;

	public static double MAX_VEL = 2;
	public static double MAX_ACC = 2.5;
	public static double MAX_JRK = 0.6;
	public static double KP = 3.0;
	public static double KI = 0.0;
	public static double KD = 0.0;
	public static double GFACT = 6.0;

	double TIME_STEP = 0.02;
	double KV = 1.0 / MAX_VEL;
	double KA = 0.0;

	double wheelbase_width = 0.86;

	static public boolean print_path = false;
	static public boolean print_trajectory = false;
	static public boolean plot_trajectory = false;
	static public boolean plot_path = true;

	static public boolean use_gyro = true;
	static public boolean debug_command = false;

	ArrayList<PathData> pathdata = new ArrayList<PathData>();
	private int pathIndex = 0;
	double runtime = 0;

	final double i2m = 0.0254;
	final double m2i = (1.0 / 0.0254);
	long start_time;

	//static double last_gyro_heading = 0;
	//static double last_target_heading = 0;

	Angle calc_angle=new Angle();
	Angle gyro_angle=new Angle();

	MyWaypoint[] points = new MyWaypoint[] { 
		new MyWaypoint(0, 0, 0), 
		new MyWaypoint(3, 0, 0),
		new MyWaypoint(13.5, -3.0, 0)
	};

	public DrivePath() {
		requires(Robot.driveTrain);
		start_time = System.nanoTime();
		config = new Trajectory.Config(Trajectory.FitMethod.HERMITE_CUBIC, Trajectory.Config.SAMPLES_FAST, TIME_STEP,
				MAX_VEL, MAX_ACC, MAX_JRK);
		trajectory = Pathfinder.generate(points, config);
		if (trajectory == null) {
			System.out.println("Uh-Oh! Trajectory could not be generated!\n");
			return;
		}

		runtime = trajectory.length() * TIME_STEP;
		// Create the Modifier Object
		TankModifier modifier = new TankModifier(trajectory);

		// Generate the Left and Right trajectories using the original trajectory
		// as the center
		modifier.modify(wheelbase_width);

		leftTrajectory = modifier.getLeftTrajectory(); // Get the Left Side
		rightTrajectory = modifier.getRightTrajectory(); // Get the Right Side

		leftFollower = new DistanceFollower(leftTrajectory);
		leftFollower.configurePIDVA(KP, KI, KD, KV, KA);
		rightFollower = new DistanceFollower(rightTrajectory);
		rightFollower.configurePIDVA(KP, KI, KD, KV, KA);
		System.out.format("trajectory length:%d runtime:%f calctime:%f\n", trajectory.length(), runtime, getProcTime());

		if (print_path) {
			double t = 0;
			for (int i = 0; i < trajectory.length(); i++) {
				Segment s = trajectory.get(i);
				System.out.format("%f %f %f %f %f %f \n", t, s.x, s.y, s.velocity, s.acceleration, s.heading);
				t += s.dt;
			}
		}
		if (print_trajectory) {
			double t = 0;
			for (int i = 0; i < trajectory.length(); i++) {
				Segment s = trajectory.get(i);
				Segment l = leftTrajectory.get(i);
				Segment r = rightTrajectory.get(i);
				System.out.format("%f %f %f %f %f\n", t, l.x, l.y, r.x, r.y);
				t += s.dt;
			}
		}
		if (plot_trajectory)
			plotCalculatedPath();
	}

	double getProcTime() {
		double curtime = 1e-9 * (System.nanoTime() - start_time);
		return (double) curtime;
	}

	double getSimTime() {
		return (double) TIME_STEP * pathIndex;
	}

	public static double feetToMeters(double feet) {
		return 2.54 * 12 * feet / 100;
	}

	// double unwrap(double previous_angle, double new_angle) {
	// 	double d = new_angle - previous_angle;
	// 	d = d >= 180 ? d - 360 : (d <= -180 ? d + 360 : d);
	// 	return previous_angle + d;
	// }

	// Called just before this Command runs the first time
	protected void initialize() {
		if (trajectory == null)
			return;
		System.out.println("DrivePath.initialize()");
		leftFollower.reset();
		rightFollower.reset();
		Robot.driveTrain.reset();
		pathIndex = 0;
		start_time = System.nanoTime();
	}

	double feet2meters(double x) {
		return 12 * x * 0.0254;
	}

	private static double metersToFeet(double meters) {
		return meters * 100 / (2.54 * 12);
	}

	// Called repeatedly when this Command is scheduled to run
	protected void execute() {
		if (trajectory == null)
			return;
		double ld = feet2meters(Robot.driveTrain.getLeftDistance());
		double rd = feet2meters(Robot.driveTrain.getRightDistance());
		double l = leftFollower.calculate(ld);
		double r = rightFollower.calculate(rd);

		double turn = 0;
		double scale = Robot.auto_scale;

		double gh = Robot.driveTrain.getHeading(); // Assuming the gyro is giving a value in degrees
		gh=gyro_angle.unwrap(gh);
		//gh = unwrap(last_gyro_heading, gh);
		//last_gyro_heading = gh;

		double th = -Pathfinder.r2d(leftFollower.getHeading()); // Should also be in degrees
		th=calc_angle.unwrap(th);
		//th = unwrap(last_target_heading, th);
		//last_target_heading = th;

		double herr = th - gh;
		if (use_gyro)
			turn = GFACT * (1.0 / 180.0) * herr;
		l += turn;
		r -= turn;
		l *= scale;
		r *= scale;
		if (debug_command)
			System.out.format("%f %f %f %f %f %f %f %f\n", getSimTime(), ld, rd, gh, th, herr, l, r);
		if (print_path || plot_path)
			debugPathError(ld, rd, l, r, gh, th);
		Robot.driveTrain.set(l, r);
		pathIndex++;
	}

	private void debugPathError(double ld, double rd, double l, double r, double g, double t) {
		PathData pd = new PathData();
		pd.tm = getSimTime();
		Segment ls = leftTrajectory.get(pathIndex);
		Segment rs = rightTrajectory.get(pathIndex);
		pd.d[0] = m2i * ld;// (Robot.driveTrain.getLeftDistance());
		pd.d[1] = m2i * (ls.position);
		pd.d[2] = m2i * rd;// (Robot.driveTrain.getRightDistance());
		pd.d[3] = m2i * (rs.position);

		// if (use_gyro) {
		pd.d[4] = g; // Assuming the gyro is giving a value in degrees
		pd.d[5] = t;
		// } else {
		// pd.d[4] = l * 10;
		// pd.d[5] = r * 10;
		// }

		if (print_path)
			System.out.format("%f %f %f %f %f\n", pd.tm, pd.d[0], pd.d[1], pd.d[2], pd.d[3]);
		if (plot_path)
			pathdata.add(pd);
	}

	// Make this return true when this Command no longer needs to run execute()
	protected boolean isFinished() {
		if (trajectory == null)
			return true;
		if (getSimTime() - runtime > 0.0)
			return true;
		if (leftFollower.isFinished() && rightFollower.isFinished())
			return true;
		return false;
	}

	// Called once after isFinished returns true
	protected void end() {
		System.out.println("DrivePath.end()");
		Robot.driveTrain.disable();
		if (plot_path)
			new PlotPath(pathdata, 0, 6, PlotPath.DIST_MODE);
	}

	// Called when another command which requires one or more of the same
	// subsystems is scheduled to run
	protected void interrupted() {
		System.out.println("DrivePath.interrupted()");
		end();
	}

	void plotCalculatedPath() {
		double time = 0;
		ArrayList<PathData> data = new ArrayList<>();
		for (int i = 0; i < trajectory.length(); i++) {
			Segment centerSegment = trajectory.get(i);
			Segment leftSegment = leftTrajectory.get(i);
			Segment rightSegment = rightTrajectory.get(i);
			double cx = metersToFeet(centerSegment.x);
			double cy = metersToFeet(centerSegment.y);
			double lx = metersToFeet(leftSegment.x);
			double ly = metersToFeet(leftSegment.y);
			double rx = metersToFeet(rightSegment.x);
			double ry = metersToFeet(rightSegment.y);

			time += centerSegment.dt;

			PathData pd = new PathData();
			// note: pathfinder inverts y order of left and right side
			pd.tm = time;
			pd.d[0] = lx;
			pd.d[1] = -ly;
			pd.d[2] = cx;
			pd.d[3] = -cy;
			pd.d[4] = rx;
			pd.d[5] = -ry;
			data.add(pd);
		}
		JFrame frame = new PlotPath(data, 0, 3, PlotPath.PATH_MODE);
		// frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frame.pack();
		frame.setLocationRelativeTo(null);
		frame.setVisible(true);
	}

	private class MyWaypoint extends Waypoint {
		MyWaypoint(double x, double y, double h) {
			super(feetToMeters(x), -feetToMeters(y), -Pathfinder.d2r(h));
		}
	};
	private class Angle {
		double previous_angle=0;
		public void reset(){
			previous_angle=0;
		}
		public double unwrap(double angle) {
			double d = angle - previous_angle;
		    d = d >= 180 ? d - 360 : (d <= -180 ? d + 360 : d);
			double new_angle=previous_angle + d;
			previous_angle=new_angle;
			return new_angle;
		}
	};
}
